x-pack/metricbeat/module/azure/billing/data.go

```diff
@@ -5,59 +5,113 @@
 package billing
 
 import (
-	"fmt"
 	"strings"
 	"time"
 
 	"github.com/Azure/azure-sdk-for-go/services/consumption/mgmt/2019-10-01/consumption"
-
 	"github.com/shopspring/decimal"
 
 	"github.com/elastic/beats/v7/metricbeat/mb"
 	"github.com/elastic/elastic-agent-libs/mapstr"
 )
 
-func EventsMapping(subscriptionId string, results Usage) []mb.Event {
+// EventsMapping maps the usage details to a slice of metricbeat events.
+func EventsMapping(subscriptionId string, results Usage, startTime time.Time, endTime time.Time) []mb.Event {
 	var events []mb.Event
 	if len(results.UsageDetails) > 0 {
-		for _, usageDetail := range results.UsageDetails {
-			event := mb.Event{
-				ModuleFields: mapstr.M{
+		for _, ud := range results.UsageDetails {
+			event := mb.Event{Timestamp: time.Now().UTC()}
+
+			// shared fields
+			event.RootFields = mapstr.M{
+				"cloud.provider": "azure",
+			}
+
+			//
+			// legacy data format
+			//
+			if legacy, isLegacy := ud.AsLegacyUsageDetail(); isLegacy {
+				event.ModuleFields = mapstr.M{
+					"subscription_id":   legacy.SubscriptionID,
+					"subscription_name": legacy.SubscriptionName,
 					"resource": mapstr.M{
-						"type":  usageDetail.ConsumedService,
-						"group": getResourceGroupFromId(*usageDetail.InstanceID),
-						"name":  usageDetail.InstanceName,
+						"name":  legacy.ResourceName,
+						"type":  legacy.ConsumedService,
+						"group": legacy.ResourceGroup,
 					},
-					"subscription_id": usageDetail.SubscriptionGUID,
-				},
-				MetricSetFields: mapstr.M{
-					"pretax_cost":       usageDetail.PretaxCost,
-					"department_name":   usageDetail.DepartmentName,
-					"product":           usageDetail.Product,
-					"usage_start":       usageDetail.UsageStart.ToTime(),
-					"usage_end":         usageDetail.UsageEnd.ToTime(),
-					"currency":          usageDetail.Currency,
-					"billing_period_id": usageDetail.BillingPeriodID,
-					"account_name":      usageDetail.AccountName,
-				},
-				Timestamp: time.Now().UTC(),
+				}
+				event.MetricSetFields = mapstr.M{
+					// original fields
+					"billing_period_id": legacy.ID,
+					"product":           legacy.Product,
+					"pretax_cost":       legacy.Cost,
+					"currency":          legacy.BillingCurrency,
+					"department_name":   legacy.InvoiceSection,
+					"account_name":      legacy.BillingAccountName,
+					"usage_start":       startTime,
+					"usage_end":         endTime,
+
+					// additional fields
+					"usage_date": legacy.Date, // Date for the usage record.
+					"account_id": legacy.BillingAccountID,
+					"unit_price": legacy.UnitPrice,
+					"quantity":   legacy.Quantity,
+				}
+				_, _ = event.RootFields.Put("cloud.region", legacy.ResourceLocation)
+				_, _ = event.RootFields.Put("cloud.instance.name", legacy.ResourceName)
+				_, _ = event.RootFields.Put("cloud.instance.id", legacy.ResourceID)
+			}
+
+			//
+			// modern data format
+			//
+			if modern, isModern := ud.AsModernUsageDetail(); isModern {
```

Comment:
should we handle the edge case (theoretically impossible) of `else if modern, isModern := ud.AsModernUsageDetail(); isModern { ... } else { return nil, fmt.Error("unsupported usage details format")  }` ?

Reply:
Better safe than sorry. 

I have updated the event mapping to return an error if the response contains no legacy or modern data.